runtime/gc_base/ObjectAccessBarrier.cpp

```diff
@@ -1528,6 +1530,9 @@ MM_ObjectAccessBarrier::copyObjectFields(J9VMThread *vmThread, J9Class *objectCl
 			/* Determine if the slot contains an object pointer or not */
 			if (descriptionBits & 1) {
 				J9Object *objectPtr = mixedObjectReadObject(vmThread, srcObject, srcOffset + offset, false);
+				if (NULL != objectMapFunction) {
```

Comment:
In general case, Read Barrier expects srcObject to be in heap, and in theory might assert if it's a stack allocated one, although I believe Concurrent Scavenger will not (it will act only if srcObj is within Evacuate Area and ignore otherwise).

All that to say, something like this might be more correct:

```
if (NULL != objectMapFunction) {
	objectPtr = objectMapFunction(vmThread, srcObject, objectMapData);
} else {
	objectPtr = mixedObjectReadObject(vmThread, srcObject, srcOffset + offset, false);
}
```

Reply:
The code is converting stack-allocated objects into heap-allocated ones, requiring us to clone the stack to the heap. The map function is required to account for stack-allocated objects pointing to other stack-allocated objects. My new code creates all of the heap objects required before cloning the contents.

I don't think the original code was incorrect, but it's much cleaner not to copy and modify the GC code, especially since I don't want to have to know the details of where/when barriers are required.

Reply:
When I said original I really meant current (but original relative to my small adjustment in an earlier comment). I don't have doubts the current approach is cleaner, but I was just not sure, if it's correct to apply a barrier on srcObj that is known to be offheap object (if mapping function exists). I guess srcObj could have happened to be a stack allocated one even before these changes, and all worked, so it should be ok now, too.

Reply:
Ah, I see the concern about the barriers. This case would not have come up before because clone() could never be called on a stack-allocated object (being the receiver of a native is certainly considered and esacpe).

I expect the JIT is able to inline clone(), but their implementation would not end up in the affected GC code.

If this turns out to be an issue, I can re-duplicate the code in the GC. I still think it's better to have the code there rather than copied into the JVMTI code.

Reply:
IIRC, the GC now uses range checks for a lot of the barrier operations, so being off-heap may be sufficient to exclude unnecessary/incorrect barriers.

Reply:
The performance of this new path (clone with a provided mapping function) is not critical.